File: src/Plugins/dotnet_albumart/AimpExtensionAlbumArtProvider.cs
//  ----------------------------------------------------
//  AIMP DotNet SDK
// 
//  Copyright (c) 2014 - 2022 Evgeniy Bogdan
//  https://github.com/martin211/aimp_dotnet
// 
//  Mail: mail4evgeniy@gmail.com
//  ----------------------------------------------------

using System.Drawing.Imaging;
using System.IO;
using AIMP.SDK.AlbumArt;
using AIMP.SDK.AlbumArt.Extensions;
using AIMP.SDK.FileManager.Objects;
using AIMP.SDK.Objects;

namespace dotnet_albumart
{
    using System.Drawing;
    using AIMP.SDK;

    public class AimpExtensionAlbumArtProvider : IAimpExtensionAlbumArtProvider
    {
        public AimpActionResult<Bitmap> Get(IAimpFileInfo fileInfo, IAimpAlbumArtSearchOptions options)
        {
            var image = Properties.Resources.front;
            return new AimpActionResult<Bitmap>(ActionResultType.OK, image);
        }

        public AimpActionResult<Bitmap> Get(string fileUrl, string artist, string album,
            IAimpAlbumArtSearchOptions options)
        {
            var image = Properties.Resources.front;
            return new AimpActionResult<Bitmap>(ActionResultType.OK, image);
        }

        public AimpAlbumArtProviderCategory GetCategory()
        {
            return AimpAlbumArtProviderCategory.Internet;
        }
    }

    internal class AimpExtensionAlbumArtProvider3 : IAimpExtensionAlbumArtProvider3
    {
        private readonly IAimpCore _core;

        public AimpExtensionAlbumArtProvider3(IAimpCore core)
        {
            _core = core;
        }

        public AimpActionResult<IAimpImageContainer> Get(IAimpFileInfo file, IAimpAlbumArtRequest requestOption)
        {
            var containerResult = _core.CreateAimpObject<IAimpImageContainer>();
            var container = containerResult.Result;

            var image = Properties.Resources.front;
            using (var stream = new MemoryStream())
            {
                image.Save(stream, ImageFormat.Jpeg);
                container.SetDataSize(stream.Length);
                var containerData = container.GetData();
                stream.Read(containerData.Result, 0, (int)stream.Length);
            }

            return new AimpActionResult<IAimpImageContainer>(ActionResultType.OK, container);
        }

        public AimpAlbumArtProviderCategory GetCategory()
        {
            return AimpAlbumArtProviderCategory.Internet;
        }
    }
}

File: src/SDK/aimp_dotnet/SDK/AlbumArt/AimpExtensionAlbumArtProvider.cpp
// ----------------------------------------------------
// AIMP DotNet SDK
// Copyright (c) 2014 - 2020 Evgeniy Bogdan
// https://github.com/martin211/aimp_dotnet
// Mail: mail4evgeniy@gmail.com
// ----------------------------------------------------

#include "Stdafx.h"
#include "AimpExtensionAlbumArtProvider.h"
#include "SDK\FileManager\AimpFileInfo.h"
#include "SDK\AlbumArt\AimpAlbumArtSearchOptions.h"

HRESULT WINAPI AimpExtensionAlbumArtProvider::Get(IAIMPString* FileURI, IAIMPString* Artist, IAIMPString* Album,
                                                  IAIMPPropertyList* Options, IAIMPImageContainer** Image) {
    IAimpAlbumArtSearchOptions^ searchOptions = gcnew AimpAlbumArtSearchOptions(Options);
    const auto r = _managedinstance->Get(
        gcnew String(FileURI->GetData()),
        gcnew String(Artist->GetData()),
        gcnew String(Album->GetData()),
        searchOptions);

    if (r->ResultType == ActionResultType::OK && r->Result != nullptr) {
        const auto container = AimpConverter::ToAimpImageContainer(r->Result);
        if (container == nullptr) {
            return E_UNEXPECTED;
        }

        *Image = container;
        return S_OK;
    }

    return E_FAIL;
}

DWORD WINAPI AimpExtensionAlbumArtProvider::GetCategory() {
    return DWORD(_managedinstance->GetCategory());
}

HRESULT WINAPI AimpExtensionAlbumArtProvider::Get2(IAIMPFileInfo* FileInfo, IAIMPPropertyList* Options,
                                                   IAIMPImageContainer** Image) {
    IAimpFileInfo^ fi = gcnew AimpFileInfo(FileInfo);
    IAimpAlbumArtSearchOptions^ searchOptions = gcnew AimpAlbumArtSearchOptions(Options);
    const auto r = ((IAimpExtensionAlbumArtProvider2^)_managedinstance)->Get(fi, searchOptions);
    if (r->ResultType == ActionResultType::OK && r->Result != nullptr) {
        const auto container = AimpConverter::ToAimpImageContainer(r->Result);
        if (container == nullptr) {
            return E_UNEXPECTED;
        }

        *Image = container;
        return S_OK;
    }

    return E_UNEXPECTED;
}
